sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java

```diff
@@ -951,15 +957,30 @@ private <T> Flux<FeedResponse<T>> createQuery(
             this.collectionCache,
             null,
             resourceLink,
-            ModelBridgeInternal.getPropertiesFromQueryRequestOptions(options));
+            ModelBridgeInternal.getPropertiesFromQueryRequestOptions(nonNullQueryOptions));
 
-        return ObservableHelper.fluxInlineIfPossibleAsObs(
-            () -> createQueryInternal(
-                resourceLink, sqlQuery, options, klass, resourceTypeEnum, queryClient, correlationActivityId, isQueryCancelledOnTimeout),
-            invalidPartitionExceptionRetryPolicy);
+
+        final ScopedDiagnosticsFactory diagnosticsFactory = new ScopedDiagnosticsFactory(this);
+        qryOptAccessor.setDiagnosticsFactoryResetCallback(nonNullQueryOptions, () -> diagnosticsFactory.reset());
+        return
+            ObservableHelper.fluxInlineIfPossibleAsObs(
+                                () -> createQueryInternal(
+                                    diagnosticsFactory, resourceLink, sqlQuery, nonNullQueryOptions, klass, resourceTypeEnum, queryClient, correlationActivityId, isQueryCancelledOnTimeout),
+                                invalidPartitionExceptionRetryPolicy
+                            )
+                            .flatMap(result -> {
+                                diagnosticsFactory.merge(nonNullRequestOptions);
```

Comment:
just for personal understanding -
I guess by theory, we do not really need the merge here any more? because now we reset for each page?

Reply:
You are amazing - yes and no - you are right that this is the wrong place for merge - causing the majority of the remaining test failures.

Reset is just Creating a new CosmosDiagnosticsContext - not merging in the CosmosDiagnostics instances - so, merge is still needed - but it should be wired up the same way as reset. Working on that right now